### PR TITLE
[RFC] feat(testlab): add docker support

### PR DESCRIPTION
**Pull request by raymondfeng**
This PR adds a few helper functions to use `docker` for tests. It's based on `dockerode` that uses REST APIs to interact with the `docker` daemon.

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

**Comment by bajtos** [comment]
I find it difficult to review this pull request because I have no idea what do you want to use the new APIs for.

Have you considered creating a new package instead of adding this functionality to `@loopback/testlab`? I am concerned that testlab is becoming too large. The initial idea was to create a thin wrapper around testing tools we use in many LB4 packages, especially to smooth out integration with TypeScript.

In my mind, Docker client is a very specialized feature that will be used only by few tests, therefore it's better to put in into a new package.

File: packages/testlab/src/__tests__/integration/docker.integration.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import {expect} from '../../expect';
+
+describe('docker', function() {
+  let client: DockerClient;
+
+  // eslint-disable-next-line no-invalid-this
+  this.timeout(30000);
+
+  before(async () => {
+    client = new DockerClient();
```
Comment (bajtos) [inline packages/testlab/src/__tests__/integration/docker.integration.ts]: How is this going to run on machines that don't have Docker daemon running? For example Windows build machines like AppVeyor? Personally, I am not running the Docker daemon unless I really need it. I would consider it as a degradation of developer experience if `npm test` in loopback-next started to require a running Docker server.

**Comment by bajtos** [comment]
@raymondfeng Have you considered to contribute this higher-level abstraction directly to `dockerode`?

File: packages/testlab/src/__tests__/integration/docker.integration.ts (hunk 1 above)
Comment (bajtos) [reply packages/testlab/src/__tests__/integration/docker.integration.ts]: Travis CI builds on MacOS are failing too.

https://travis-ci.com/strongloop/loopback-next/jobs/245526182
```
 1) docker
       "before all" hook for "inspects an image by name":
     Error: connect ENOENT /var/run/docker.sock
      at PipeConnectWrap.afterConnect [as oncomplete] (net.js:1128:14)
```

**Comment by raymondfeng** [comment]
@bajtos Thank you for the feedback. 

I see the need to have simple Docker APIs that allow us to pull/start a Docker container for popular external resources such as Redis, MongoDB, and Kafka for integration/acceptance tests while I'm experimenting Cloud Deployment and Pub/Sub. 

I see your points that it does not fit `testlab` well. I'll try to find a different avenue for it.

- Contribute to dockerode
- Create a separate package

**Comment by raymondfeng** [comment]
@bajtos Does it make sense to have docker-related helpers to be part of our `@loopback/build` module?

**Comment by raymondfeng** [comment]
I just discovered https://github.com/testcontainers/testcontainers-node. It seems to a good fit for my use cases.

**Pull request closed without merging**